apps/src/codebridge/PopUpButton/PopUpButton.tsx

```diff
@@ -0,0 +1,72 @@
+import React, {useState, useCallback} from 'react';
+
+import Button from '@cdo/apps/componentLibrary/button';
+
+import moduleStyles from './PopUpButton.module.scss';
+
+type PopUpButtonProps = {
+  iconName: string;
+  children?: React.ReactNode;
+  className?: string;
+  alignment?: 'left' | 'right';
+};
+
+export const PopUpButton = ({
+  children,
+  iconName,
+  className,
+  alignment = 'left',
+}: PopUpButtonProps) => {
+  const [isOpen, _setIsOpen] = useState(false);
+  const [buttonRect, setButtonRect] = useState<DOMRect>();
+  const [offsetParent, setOffsetParent] = useState<DOMRect>();
+
+  const setIsOpenFalse = useCallback(() => {
+    _setIsOpen(false);
+    document.removeEventListener('click', setIsOpenFalse);
+  }, [_setIsOpen]);
+
+  const setIsOpen = useCallback(
+    (newIsOpen: boolean) => {
+      _setIsOpen(newIsOpen);
+      if (newIsOpen) {
+        document.addEventListener('click', setIsOpenFalse);
+      } else {
+        document.removeEventListener('click', setIsOpenFalse);
+      }
+    },
+    [_setIsOpen, setIsOpenFalse]
+  );
+
+  return (
+    <>
+      <Button
+        className={className}
+        size="xs"
+        icon={{iconStyle: 'solid', iconName}}
+        color="black"
+        isIconOnly
+        onClick={e => {
+          setButtonRect((e.target as HTMLElement).getBoundingClientRect());
```

Comment:
nit: extract this out into a helper method?

Reply:
done, and also re-tooled and fiddled with the logic to help address the previous comment about the three methods, getting rid of the `_setIsOpen`/`setIsOpen` wrapper method.

There may be different ways to implement all of this, so take it with a grain of salt as a personal pref. The nuisance was the event listener which was tacked onto the document. _That_ event listener is the thing that responds to clicks elsewhere in the doc to close our pop up menu, and we want to clean it up when we no longer need it. In this case, that's when the pop up menu closes. Annoyingly, the only way we can remove an event listener is by handing in the same function we'd originally added.

So we need to keep a stable copy of that function, which is the `setIsOpenFalse` that does nothing more than set `isOpen` to false, as well as removing itself as the click handler. I'd originally added the additional `setIsOpen` wrapper as a way of keeping the logic for adding/removing the click handler closer to changing the `isOpen` value itself, as opposed to associating it more closely with the click handler. But that's personal preference and it could've gone in the other place.

...which is where I ended up moving it anyway. When I pulled out the embedded callback to stick in a separate helper with a `useCallback`, I lost access to the current `isOpen` value (w/o making it a callback dependency and making `useCallback` moot), and at that point I needed to move the event listener adding/removing into the `setState` callback anyway (to get at the `newIsOpen` value). Anyway, `setState` function is now effectively the old wrapper `setIsOpen` that set the value + added/removed the listeners, so it's in place there now; just embedded into the overall click handler.

`setIsOpenFalse` is admittedly possibly too short as a name, but conversely I felt that `setIsOpenFalseAndRemoveEventListener` was too long. Definitely open to better suggestions on it, though.

As I've often repeated, there are only 2 difficult tasks in computer science: Cache invalidation, naming things, and off by one errors.

Reply:
thanks for the explanation!